### PR TITLE
Implementation of InvokeWMLMethod

### PR DESCRIPTION
**Pull request by sedgewickmm18**
added a function to bif.py

### CODE REVIEW

File: iotfunctions/bif.py [hunk 1]
```diff
@@ -2816,6 +2816,166 @@ def execute(self, df):
...
+                pass
+            self.deployment_id = wml_credentials['deployment_id']
+            self.space_id = wml_credentials['space_id']
+        else:
+            wml_credentials = {'apikey': self.apikey , 'url': self.wml_endpoint, 'space_id': self.space_id}
+
+        # get client and check credentials
+        self.client = APIClient(wml_credentials)
+        if self.client is None:
+            logger.error('WML API Key invalid')
```
Comment (pkohlmann) [inline iotfunctions/bif.py:2899]: logger.error() in conjunction with 'raise Exception' leaves the stacktrace twice in logfile what is umfavourable. I would omit the logger.error() statement.

File: iotfunctions/bif.py [hunk 2]
```diff
@@ -2816,6 +2816,166 @@ def execute(self, df):
...
+        if self.logged_on:
+            return
+
+        # retrieve WML credentials as constant
+        #    {"apikey": api_key, "url": 'https://' + location + '.ml.cloud.ibm.com'}
+        if self.wml_auth is not None:
+            c = self._entity_type.get_attributes_dict()
+            try:
+                wml_credentials = c[self.wml_auth]
+                print('WML Credentials ' , str(wml_credentials))
```
Comment (pkohlmann) [inline iotfunctions/bif.py:2886]: Remove print statement or replace it by logger.info/debug(). In this case, the log message contains credentials which are not supposed to show up in logfiles. Therefore remove line completely or obfuscate the credentials

File: iotfunctions/bif.py [hunk 3]
```diff
@@ -2816,6 +2816,166 @@ def execute(self, df):
...
+            except Exception as ae:
+                wml_credentials = {'apikey': self.apikey , 'url': self.wml_endpoint, 'space_id': self.space_id}
+                logger.error('WML Credentials constant ' + self.wml_auth + ' not present. Error ' + str(ae))
+                pass
```
Comment (pkohlmann) [inline iotfunctions/bif.py:2890]: You dump an error with stacktrace into logfile by logger.error() but you continue execution. I guess this should be a warning that there is an attempt to circumvent the missing constant. By the way, the 'pass' is not necessary.

File: iotfunctions/bif.py [hunk 4]
```diff
@@ -2816,6 +2816,166 @@ def execute(self, df):
...
+    def __str__(self):
+        out = self.__class__.__name__
+        try:
+            out = out + 'Input: ' + str(self.input_items) + '\n'
+            out = out + 'Output: ' + str(self.output_items) + '\n'
+
+            if self.wml_auth is not None:
+                out = out + 'WML auth: ' + str(self.wml_auth) + '\n'
+            else:
+                out = out + 'APIKey: ' + str(self.apikey) + '\n'
```
Comment (pkohlmann) [inline iotfunctions/bif.py:2865]: Do you really want to make the APIKey visible?

**Review by pkohlmann — CHANGES_REQUESTED** [review]

File: iotfunctions/bif.py [hunk 5]
```diff
@@ -2816,6 +2816,166 @@ def execute(self, df):
...
+        self.login()
+
+        return super().execute(df)
+
+    def _calc(self, df):
+
+        if (len(self.input_items) == 1):
+            logging.debug('reformating column ' + str(self.input_items))
+            s_df = df[self.input_items]
+            rows = [list(r) for i,r in s_df.iterrows()]
```
[the remark this answers is not in the document]
Comment (sedgewickmm18) [reply iotfunctions/bif.py]: Using tolist() now

**Review by pkohlmann — APPROVED** [review]
All good now!

**Pull request merged**